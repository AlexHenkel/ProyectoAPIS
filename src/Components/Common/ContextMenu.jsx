import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import IconButton from 'material-ui/IconButton'
import Menu, { MenuItem } from 'material-ui/Menu'
import MoreVertIcon from 'material-ui-icons/MoreVert'
import EditIcon from 'material-ui-icons/Edit'
import RemoveIcon from 'material-ui-icons/Delete'
import { ListItemIcon } from 'material-ui'

export const ContextContainer = styled.div`
  position: absolute;
  right: 0;
  top: 7px;
`

class ContextMenu extends Component {
  constructor(props) {
    super(props)
    this.state = {
      anchorEl: null,
      open: false,
    }
  }

  handleClick = (event) => {
    this.setState({ open: true, anchorEl: event.currentTarget })
  }

  handleRequestClose = () => {
    this.setState({ open: false })
  }

  handleClickItem = cb => () => {
    const { value } = this.props
    cb(value)
    this.handleRequestClose()
  }

  render() {
    const { anchorEl, open } = this.state
    const { handleEdit, handleRemove } = this.props
    return (
      <div>
        <IconButton
          aria-label="More"
          aria-owns={this.state.open ? 'long-menu' : null}
          aria-haspopup="true"
          onClick={this.handleClick}
        >
          <MoreVertIcon />
        </IconButton>
        <Menu
          anchorEl={anchorEl}
          open={open}
          onRequestClose={this.handleRequestClose}
        >
          <MenuItem onClick={this.handleClickItem(handleEdit)}>
            <ListItemIcon>
              <EditIcon />
            </ListItemIcon>
            Editar
          </MenuItem>
          <MenuItem onClick={this.handleClickItem(handleRemove)}>
            <ListItemIcon>
              <RemoveIcon />
            </ListItemIcon>
            Eliminar
          </MenuItem>
        </Menu>
      </div>
    )
  }
}

ContextMenu.propTypes = {
  handleEdit: PropTypes.func.isRequired,
  handleRemove: PropTypes.func.isRequired,
  value: PropTypes.any.isRequired,
}

export default ContextMenu
