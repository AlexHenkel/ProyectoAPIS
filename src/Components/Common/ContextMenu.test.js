import React from 'react'
import ContextMenu from './ContextMenu'

describe('Context menu component', () => {
  let component

  beforeEach(() => {
    component = (
      <ContextMenu
        handleEdit={jest.fn()}
        handleRemove={jest.fn()}
        value={1}
      />
    )
  })

  it('renders correctly', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render prop', () => {
    const contextMenu = mount(component)
    expect(contextMenu.props().handleEdit).toBeDefined()
    expect(contextMenu.props().handleRemove).toBeDefined()
    expect(contextMenu.props().value).toBeDefined()
  })

  it('should open on click and closed on requestClose', () => {
    const contextMenu = shallow(component)
    expect(contextMenu).toMatchSnapshot()
    contextMenu.find('withStyles(IconButton)').simulate('click', { currentTarget: <div /> })
    expect(contextMenu.state().open).toEqual(true)
    contextMenu.instance().handleRequestClose()
    expect(contextMenu.state().open).toEqual(false)
  })

  it('should pass value to handleEdit on click in first context menu item', () => {
    const handleEdit = jest.fn()
    const value = 1
    const customComponent = (
      <ContextMenu handleEdit={handleEdit} handleRemove={jest.fn()} value={value} />
    )
    const contextMenu = shallow(customComponent)
    expect(contextMenu).toMatchSnapshot()
    contextMenu.find('withStyles(IconButton)').simulate('click', { currentTarget: <div /> })
    expect(contextMenu).toMatchSnapshot()
    contextMenu.find('withStyles(MenuItem)').first().simulate('click')
    expect(handleEdit).toBeCalledWith(value)
  })

  it('should pass value to handleRemove on click in last context menu item', () => {
    const handleRemove = jest.fn()
    const value = 1
    const customComponent = (
      <ContextMenu handleEdit={jest.fn()} handleRemove={handleRemove} value={value} />
    )
    const contextMenu = shallow(customComponent)
    expect(contextMenu).toMatchSnapshot()
    contextMenu.find('withStyles(IconButton)').simulate('click', { currentTarget: <div /> })
    expect(contextMenu).toMatchSnapshot()
    contextMenu.find('withStyles(MenuItem)').last().simulate('click')
    expect(handleRemove).toBeCalledWith(value)
  })
})
