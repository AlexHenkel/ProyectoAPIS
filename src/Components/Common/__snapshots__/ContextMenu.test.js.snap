// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context menu component renders correctly 1`] = `
<div>
  <button
    aria-haspopup="true"
    aria-label="More"
    aria-owns={null}
    className="MuiButtonBase-root-10 MuiIconButton-root-1"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label-7"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-12 MuiIconButton-icon-8"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root-13"
    />
  </button>
</div>
`;

exports[`Context menu component should open on click and closed on requestClose 1`] = `
<div>
  <withStyles(IconButton)
    aria-haspopup="true"
    aria-label="More"
    aria-owns={null}
    onClick={[Function]}
  >
    <pure(MoreVert) />
  </withStyles(IconButton)>
  <withStyles(Menu)
    anchorEl={null}
    onRequestClose={[Function]}
    open={false}
  >
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Edit) />
      </withStyles(ListItemIcon)>
      Editar
    </withStyles(MenuItem)>
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Delete) />
      </withStyles(ListItemIcon)>
      Eliminar
    </withStyles(MenuItem)>
  </withStyles(Menu)>
</div>
`;

exports[`Context menu component should pass value to handleEdit on click in first context menu item 1`] = `
<div>
  <withStyles(IconButton)
    aria-haspopup="true"
    aria-label="More"
    aria-owns={null}
    onClick={[Function]}
  >
    <pure(MoreVert) />
  </withStyles(IconButton)>
  <withStyles(Menu)
    anchorEl={null}
    onRequestClose={[Function]}
    open={false}
  >
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Edit) />
      </withStyles(ListItemIcon)>
      Editar
    </withStyles(MenuItem)>
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Delete) />
      </withStyles(ListItemIcon)>
      Eliminar
    </withStyles(MenuItem)>
  </withStyles(Menu)>
</div>
`;

exports[`Context menu component should pass value to handleEdit on click in first context menu item 2`] = `
<div>
  <withStyles(IconButton)
    aria-haspopup="true"
    aria-label="More"
    aria-owns="long-menu"
    onClick={[Function]}
  >
    <pure(MoreVert) />
  </withStyles(IconButton)>
  <withStyles(Menu)
    anchorEl={<div />}
    onRequestClose={[Function]}
    open={true}
  >
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Edit) />
      </withStyles(ListItemIcon)>
      Editar
    </withStyles(MenuItem)>
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Delete) />
      </withStyles(ListItemIcon)>
      Eliminar
    </withStyles(MenuItem)>
  </withStyles(Menu)>
</div>
`;

exports[`Context menu component should pass value to handleRemove on click in last context menu item 1`] = `
<div>
  <withStyles(IconButton)
    aria-haspopup="true"
    aria-label="More"
    aria-owns={null}
    onClick={[Function]}
  >
    <pure(MoreVert) />
  </withStyles(IconButton)>
  <withStyles(Menu)
    anchorEl={null}
    onRequestClose={[Function]}
    open={false}
  >
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Edit) />
      </withStyles(ListItemIcon)>
      Editar
    </withStyles(MenuItem)>
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Delete) />
      </withStyles(ListItemIcon)>
      Eliminar
    </withStyles(MenuItem)>
  </withStyles(Menu)>
</div>
`;

exports[`Context menu component should pass value to handleRemove on click in last context menu item 2`] = `
<div>
  <withStyles(IconButton)
    aria-haspopup="true"
    aria-label="More"
    aria-owns="long-menu"
    onClick={[Function]}
  >
    <pure(MoreVert) />
  </withStyles(IconButton)>
  <withStyles(Menu)
    anchorEl={<div />}
    onRequestClose={[Function]}
    open={true}
  >
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Edit) />
      </withStyles(ListItemIcon)>
      Editar
    </withStyles(MenuItem)>
    <withStyles(MenuItem)
      onClick={[Function]}
    >
      <withStyles(ListItemIcon)>
        <pure(Delete) />
      </withStyles(ListItemIcon)>
      Eliminar
    </withStyles(MenuItem)>
  </withStyles(Menu)>
</div>
`;
